### PR TITLE
Hooks for device/user ek generation

### PR DESCRIPTION
**Pull request by joshblum**
depends on #10914 and #10923 

We generation keys once a day and check during login/hourly if we need to regenerate. during these checks we also cleanup any stale keys.

### CODE REVIEW

File: go/ephemeral/common.go [hunk 1]
```diff
@@ -10,7 +10,10 @@ import (
 	"github.com/keybase/client/go/protocol/keybase1"
 )
 
+// Keys last at most one week
 const KeyLifetimeSecs = keybase1.Time(time.Hour * 24 * 7) // one week
+// Everyday we want to generate a new key if possible
+const KeyGenLifetimeSecs = keybase1.Time(time.Hour * 24) // one day
```
Comment (oconnor663) [inline go/ephemeral/common.go]: Maybe preexisting, are these units right? I think time.Hour is nanos.

File: go/service/main.go [hunk 2]
```diff
@@ -554,6 +554,9 @@ func (d *Service) hourlyChecks() {
 			d.G().Log.Debug("+ hourly check loop")
 			d.G().Log.Debug("| checking tracks on an hour timer")
 			libkb.CheckTracking(d.G())
+			d.G().Log.Debug("| checking if ephemeral keys need to be created or deleted")
+			ekLib := d.G().GetEKLib()
+			ekLib.KeygenIfNeeded(context.Background())
```
Comment (oconnor663) [inline go/service/main.go]: It looks like this happens an hour after startup. Should we be more aggressive?

File: go/ephemeral/common.go (hunk 1 above)
Comment (joshblum) [reply go/ephemeral/common.go]: ah yes, fixed

File: go/service/main.go (hunk 2 above)
Comment (joshblum) [reply go/service/main.go]: yeah i'll call it again above so it kicks off when the service starts (as long as the user is logged in)

File: go/ephemeral/common.go [hunk 3]
```diff
@@ -4,13 +4,21 @@ import (
...
 	"sort"
-	"time"
 
 	"github.com/keybase/client/go/libkb"
 	"github.com/keybase/client/go/protocol/keybase1"
 )
 
-const KeyLifetimeSecs = keybase1.Time(time.Hour * 24 * 7) // one week
+// Keys last at most one week
+const KeyLifetimeSecs = keybase1.Time(60 * 60 * 24 * 7) // one week
```
Comment (oconnor663) [inline go/ephemeral/common.go]: I think these units are still off. The `keybase1.Time` type is an `int64` of milliseconds, but the number we're casting is in seconds. I think in general we should avoid doing a cast, and try to use the conversion functions. If that makes it impossible to have a `const`, then we could consider exposing helper constants from inside of `keybase1`, similar to how the `time` module does?

File: go/ephemeral/common_test.go [hunk 4]
```diff
@@ -67,7 +67,7 @@ func TestDeleteExpiredKeys(t *testing.T) {
 
 	// Test with a 6 day gap, but no expiry
 	keyMap = keyExpiryMap{
-		0: now - keybase1.Time(time.Hour*24*6),
+		0: now - keybase1.Time(60*60*24*6),
```
Comment (oconnor663) [inline go/ephemeral/common_test.go]: Ditto.

File: go/ephemeral/lib.go [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+	if err != nil {
+		return err
+	}
+
+	if userEKNeeded {
+		_, err = publishNewUserEK(ctx, e.G(), merkleRoot)
+		if err != nil {
+			return err
+		}
+	}
```
Comment (oconnor663) [inline go/ephemeral/lib.go]: Can we add some TODO's here about making the device and user uploads atomic? We did that whiteboard example about how you can get into trouble if you publish a new deviceEK but fail to publish a new userEK. The short version was that the new deviceEK cuts off the extension window for the old key (by design), but without a new userEK you still have new teamEKs being created that only the old deviceEK can ultimately decrypt.

File: go/ephemeral/common.go (hunk 3 above)
Comment (joshblum) [reply go/ephemeral/common.go]: off everywhere haha. fixing

File: go/ephemeral/lib.go (hunk 5 above)
Comment (joshblum) [reply go/ephemeral/lib.go]: left a note to myself to discuss further in person

**Pull request merged**